components/feature/session/src/main/java/mozilla/components/feature/session/TrackingProtectionUseCases.kt

```diff
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.session
+
+import mozilla.components.browser.session.Session
+import mozilla.components.browser.session.SessionManager
+import mozilla.components.concept.engine.Engine
+import mozilla.components.concept.engine.content.blocking.TrackerLog
+
+/**
+ * Contains use cases related to the tracking protection.
+ *
+ * @param sessionManager the application's [SessionManager].
+ * @param engine the application's [Engine].
+ */
+class TrackingProtectionUseCases(
+    val sessionManager: SessionManager,
+    val engine: Engine
+) {
+
+    class FetchTrackingLogUserCase internal constructor(
+        private val sessionManager: SessionManager,
+        private val engine: Engine
+    ) {
+        /**
+         * Fetch all the tracking protection logged information of a given [session].
+         *
+         * @param session the session for which loading should be stopped.
+         * @param onSuccess callback invoked if the data was fetched successfully.
+         * @param onError (optional) callback invoked if fetching the data caused an exception.
+         */
+        operator fun invoke(
+            session: Session,
+            onSuccess: (List<TrackerLog>) -> Unit,
+            onError: (Throwable) -> Unit = {}
+        ) {
+            val engineSession = requireNotNull(sessionManager.getEngineSession(session))
```

Comment:
The require may throw if an app requests trackers when the engine session doesn't exist. For example, in Fenix, that could happen by trying to request trackers on the homescreen.

`getOrCreateEngineSession` might suffice.

Reply:
Improved!